Cohere provider tutorial (#3091)
astronomer/docs: learn/airflow-cohere.md

Cohere provides highly specialized out-of-the box and custom LLMs. These models are used in countless applications, both user-facing, such as to moderate user-generated content, and internal, for example providing insight into customer support tickets.

Integrating Cohere with Airflow into one end-to-end machine learning pipeline allows you to:

- Use Airflow's [data-driven scheduling](airflow-datasets.md) to run operations using Cohere LLM endpoints based on upstream events in your data ecosystem, such as when new user input is ingested or a new dataset is available.
- Send several requests to a model endpoint in parallel based on upstream events in your data ecosystem or user input via [Airflow params](airflow-params.md).
- Add Airflow features like [retries](rerunning-dags.md#automatically-retry-tasks) and [alerts](error-notifications-in-airflow.md) to your Cohere operations. This is critical for day 2 MLOps operations, for example to handle model service outages.
- Use Airflow to orchestrate the creation of vector embeddings using Cohere models, which is especially useful for very large datasets that cannot be processed automatically by vector databases.

## Time to complete

> This tutorial takes approximately 15 minutes to complete excluding optional steps.

Reviewer: ```suggestion
This tutorial takes approximately 15 minutes to complete (cooking your recommended recipe not included).
```

## Assumed knowledge

To get the most out of this tutorial, make sure you have an understanding of:

> - The basics of the Cohere API. See [Cohere Documentation](https://docs.cohere.com/reference/about).

Reviewer: Is knowledge of the Cohere API enough to ensure knowledge of what a vector embedding is, and other ML topics related to this example? If not, do you think it makes sense to add some sort of ML basics resource here?

Author: Good point! I think this is also relevant with the OpenAI tutorial for people who might only be familiar with the generation endpoint. Will add that to both (both tools have a guide on embeddings).
I think that should be enough, unless we want to have PCA as a prerequisite but you don't really need to understand that to run the example 🤔

    ```text
    apache-airflow-providers-cohere==1.0.0
    matplotlib==3.8.1
    seaborn==0.13.0
    scikit-learn==1.3.2
    pandas==1.5.3
    numpy==1.26.2
    adjustText==0.8
    ```

3. To create an [Airflow connection](connections.md) to Cohere, add the following environment variables to your `.env` file. Make sure to provide you own Cohere API key.

    ```text
    AIRFLOW_CONN_COHERE_DEFAULT='{
        "conn_type": "cohere",
        "password": "YOUR COHERE API KEY",
    }'
    ```

## Step 2: Create your DAG

1. In your `dags` folder, create a file called `recipe_suggestions.py`.

2. Copy the following code into the file.

    <CodeBlock language="python">{recipe_suggestions}</CodeBlock>

    This DAG consists of five tasks to make a simple MLOps pipeline.

    ![Screenshot of the Airflow UI showing the successful completion of the `recipe_suggestions` DAG in the Grid view with the Graph tab selected. 6 countries were provided to get recipes suggestions from, which led to 8 mapped task instances of both the `get_a_recipe` and `get_embeddings` task.](/img/tutorials/airflow-cohere_dag_graph.png)

>     - The `get_ingredients` task fetches the list of ingredients that the user found in their pantry and wants to use in their recipe. The input `pantry_ingredients` is provided via [Airflow params](airflow-params.md).

Reviewer: ```suggestion
    - The `get_ingredients` task fetches the list of ingredients that the user found in their pantry and wants to use in their recipe. The input `pantry_ingredients` param is provided via [Airflow params](airflow-params.md) when you run the DAG.
```

    - The `get_countries` task retrieves the list of user-provided countries to get recipes from via [Airflow params](airflow-params.md).

>     - The `get_a_recipe` task uses the [CohereHook](https://airflow.apache.org/docs/apache-airflow-providers-cohere/stable/_api/airflow/providers/cohere/hooks/cohere/index.html) to connect to the Cohere API and use the [`/generate` endpoint](https://docs.cohere.com/reference/generate) to get a tasty recipe suggestion for the user's pantry ingredients and one of the countries they provided. This task is [dynamically mapped](dynamic-tasks.md) over the list of countries to generate one dynamically mapped task instance per country. The recipes are saved as `.txt` files in the `include` folder.

Reviewer: ```suggestion
    - The `get_a_recipe` task uses the [CohereHook](https://airflow.apache.org/docs/apache-airflow-providers-cohere/stable/_api/airflow/providers/cohere/hooks/cohere/index.html) to connect to the Cohere API and use the [`/generate` endpoint](https://docs.cohere.com/reference/generate) to get a tasty recipe suggestion based on the user's pantry ingredients and one of the countries they provided. This task is [dynamically mapped](dynamic-tasks.md) over the list of countries to generate one task instance per country. The recipes are saved as `.txt` files in the `include` folder.
```

    - `type`: Select your preferred recipe type. 
    - `max_tokens_recipe`: The maximum number of tokens available for the recipe. 
    - `randomness_of_recipe`: The randomness of the recipe. The value provided is divided by 10 and given to the [`temperature` parameter](https://docs.cohere.com/docs/temperature) of the of the Cohere API. The scale for the param ranges from 0 to 50, with 0 being the most deterministic and 50 being the most random.

    ![Screenshot of the Airflow UI showing the params available for the `recipe_suggestions` DAG with the default choices.](/img/tutorials/airflow-cohere_params.png)

3. Go to the `include` folder to view the image file created by the `plot_embeddings` task. The image should look similar to the one below.

    ![Screenshot of the image created by the `plot_embeddings` task showing the two dimensional representation of the closeness of recipes associated with different countries.](/img/tutorials/airflow-cohere_recipe_plot.png)

## Step 4: (Optional) Cook your recipe

1. Choose one of the recipes in the `include` folder.

> 2. Navigate to your kitchen and cook the recipe you generated using Cohere with Airflow.

Reviewer: This makes me very happy :)

---
title: "Orchestrate Cohere LLMs with Apache Airflow"
sidebar_label: "Cohere"
description: "Learn how to integrate Cohere and Airflow."
id: airflow-cohere
sidebar_custom_props: { icon: 'img/integrations/cohere.png' }
---

import CodeBlock from '@theme/CodeBlock';
import recipe_suggestions from '!!raw-loader!../code-samples/dags/airflow-cohere/recipe_suggestions.py';

> [Cohere](https://cohere.com/) is a platform geared towards implementation of language AI that provides an API for accessing cutting edge large language models (LLMs). The [Cohere Airflow provider](https://airflow.apache.org/docs/apache-airflow-providers-cohere/stable/index.html) offers modules to easily integrate Cohere with Airflow.

Reviewer 2: ```suggestion
[Cohere](https://cohere.com/) is a platform where you can implement language AI that provides an API to access cutting-edge large language models (LLMs). The [Cohere Airflow provider](https://airflow.apache.org/docs/apache-airflow-providers-cohere/stable/index.html) offers modules to easily integrate Cohere with Airflow.
```

Author: Hmm... with the rephrasing I am not sure if `a platform where you can implement language AI` is the best way to describe Cohere, I guess it depends on the understanding of the word "implement". 

How about:

```
[Cohere](https://cohere.com/) is a natural language processing (NLP) platform that 
provides to access cutting-edge large language models (LLMs).
```

? Since they also have a text classification endpoint :)

> In this tutorial you'll use Airflow and the Cohere Airflow provider to generate recipe suggestions based on a list of ingredients and countries of recipe origin. You'll use Cohere to create embeddings of the recipes and dimensionality reduction to plot recipe similarity in two dimensions.

Reviewer 2: ```suggestion
In this tutorial, you use Airflow and the Cohere Airflow provider to generate recipe suggestions based on a list of ingredients and countries of recipe origin. You create embeddings of the recipes and dimensionality reduction to plot recipe similarity in two dimensions with Cohere's tools.
```
or something like that - just to break up sentence structure so we don't have multiple "you use _" structures back to back.

## Why use Airflow with Cohere?

> Cohere provides highly specialized out-of-the box and custom LLMs. These models are used in countless applications, both user-facing, such as to moderate user-generated content, and internal, for example providing insight into customer support tickets.

Reviewer 2: ```suggestion
Cohere provides highly specialized out-of-the box and custom LLMs.  Countless applications use these models for both user-facing needs, such as to moderate user-generated content, and internal purposes, like providing insight into customer support tickets.
```

> Cohere provides highly specialized out-of-the box and custom LLMs.  Countless applications use these models for both user-facing needs, such as to moderate user-generated content, and internal purposes, like providing insight into customer support tickets.

Reviewer 2: ```suggestion
Cohere provides highly specialized out-of-the box and custom LLMs. Countless applications use these models for both user-facing needs, such as to moderate user-generated content, and internal purposes, like providing insight into customer support tickets.
```

Integrating Cohere with Airflow into one end-to-end machine learning pipeline allows you to:

> - Use Airflow's [data-driven scheduling](airflow-datasets.md) to run operations using Cohere LLM endpoints based on upstream events in your data ecosystem, such as when new user input is ingested or a new dataset is available.

Reviewer 2: ```suggestion
- Use Airflow's [data-driven scheduling](airflow-datasets.md) to run operations with Cohere LLM endpoints based on upstream events in your data ecosystem, like when new user input is ingested or a new dataset is available.
```

> - Send several requests to a model endpoint in parallel based on upstream events in your data ecosystem or user input via [Airflow params](airflow-params.md).

Reviewer 2: ```suggestion
- Send several requests to a model endpoint in parallel based on upstream events in your data ecosystem or user input with [Airflow params](airflow-params.md).
```

Author: And there goes another dollar into the "via" jar 😳

> - Add Airflow features like [retries](rerunning-dags.md#automatically-retry-tasks) and [alerts](error-notifications-in-airflow.md) to your Cohere operations. This is critical for day 2 MLOps operations, for example to handle model service outages.

Reviewer 2: ```suggestion
- Add Airflow features like [retries](rerunning-dags.md#automatically-retry-tasks) and [alerts](error-notifications-in-airflow.md) to your Cohere operations. This is critical for day 2 MLOps operations, for example, to handle model service outages.
```

> - Use Airflow to orchestrate the creation of vector embeddings using Cohere models, which is especially useful for very large datasets that cannot be processed automatically by vector databases.

Reviewer 2: ```suggestion
- Use Airflow to orchestrate the creation of vector embeddings with Cohere models, which is especially useful for very large datasets that cannot be processed automatically by vector databases.
```

## Step 1: Configure your Astro project

1. Create a new Astro project:

    ```sh
    $ mkdir astro-cohere-tutorial && cd astro-cohere-tutorial
    $ astro dev init
    ```

2. Add the following lines to your `requirements.txt` file to install the Cohere Airflow provider and other supporting packages:

    ```text
    apache-airflow-providers-cohere==1.0.0
    matplotlib==3.8.1
    seaborn==0.13.0
    scikit-learn==1.3.2
    pandas==1.5.3
    numpy==1.26.2
    adjustText==0.8
    ```

3. To create an [Airflow connection](connections.md) to Cohere, add the following environment variables to your `.env` file. Make sure to provide you own Cohere API key.

    ```text
    AIRFLOW_CONN_COHERE_DEFAULT='{
        "conn_type": "cohere",
        "password": "YOUR COHERE API KEY",
    }'
    ```

## Step 2: Create your DAG

1. In your `dags` folder, create a file called `recipe_suggestions.py`.

2. Copy the following code into the file.

    <CodeBlock language="python">{recipe_suggestions}</CodeBlock>

    This DAG consists of five tasks to make a simple MLOps pipeline.

>     - The `get_ingredients` task fetches the list of ingredients that the user found in their pantry and wants to use in their recipe. The input `pantry_ingredients` param is provided via [Airflow params](airflow-params.md) when you run the DAG.

Reviewer 2: ```suggestion
    - The `get_ingredients` task fetches the list of ingredients that the user found in their pantry and wants to use in their recipe. The input `pantry_ingredients` param is provided by [Airflow params](airflow-params.md) when you run the DAG.
```

Author: 💵

>     - The `get_countries` task retrieves the list of user-provided countries to get recipes from via [Airflow params](airflow-params.md).

Reviewer 2: ```suggestion
    - The `get_countries` task uses [Airflow params](airflow-params.md) to retrieve the list of user-provided countries to get recipes from.
```
Lots of nested prepositional phrases made this sentence a little difficult to parse.

Author: 😅 yeah, overdid it here. Also 💵

    - The `plot_embeddings` task takes the embeddings created by the upstream task and performs dimensionality reduction using [PCA](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html) to plot the embeddings in two dimensions. 

    ![Screenshot of the Airflow UI showing the successful completion of the `recipe_suggestions` DAG in the Grid view with the Graph tab selected. 6 countries were provided to get recipes suggestions from, which led to 8 mapped task instances of both the `get_a_recipe` and `get_embeddings` task.](/img/tutorials/airflow-cohere_dag_graph.png)

## Step 3: Run your DAG

1. Run `astro dev start` in your Astro project to start Airflow and open the Airflow UI at `localhost:8080`.

2. In the Airflow UI, run the `recipe_suggestions` DAG by clicking the play button. Then, provide [Airflow params](airflow-params.md) for:

>     - `Countries of recipe origin`: A list of the countries you want to get recipe suggestions from. Make sure to create one line per country and to provide at least two countries.

Reviewer 2: Are there any formatting, spelling or other requirements for this? For example, does it need to be the English name of the country (Deutschland vs Germany)

Author: Good callout! No, there are no restrictions. 
The input is injected into the LLM prompt which should be able to handle any language. I just tried it and Schweiz and even Schwiiz worked. It also works with cities or continents, though `Vulcan` was disappointing, it did say "Vulcan twist" but gave a recipe from England just called it `Vulcan Olive and Gruyere Potato Pie` 😅
